### PR TITLE
Add ThreadingOptions to C#

### PR DESCRIPTION
**Pull request by thoron**
## Add new ThreadingOptions class to ORT for C#. 

Add new method for creating OrtEnv via `GetEnvironment(ThreadingOptions threadingOptions)`.

### The SetGlobalSpinControl() method in c_api takes int as `allow_spinning` 
 ```c
OrtStatus * OrtApi::SetGlobalSpinControl(OrtThreadingOptions * tp_options, int allow_spinning)
````
where `allow_spinning` is:
> 0 = It won't spin (recommended if CPU usage is high)"
> 1 = Threadpool will spin to wait for queue to become non-empty"

And I added to the C# wrapper: ` int allowSpinning` switching bool to int as `value ? 1 : 0`. Is this necessary or
will the marshalling handle bool to int?

https://github.com/microsoft/onnxruntime/blob/main/include/onnxruntime/core/session/onnxruntime_c_api.h#L2410

### How to add OrtEnv to InferenceSession ctor
Here I just added OrtEnv as a parameter to a single ctor to be able to test.

https://github.com/thoron/onnxruntime/blob/main/csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.shared.cs#L146

I want to refactor the all the constructors to have the same pattern as the Init method:

```csharp
Init(string modelPath, SessionOptions options,
                          PrePackedWeightsContainer prepackedWeightsContainer = null, IntPtr? envHandle = null)
```
```csharp
Init(byte[] modelData, SessionOptions options,
                          PrePackedWeightsContainer prepackedWeightsContainer = null, IntPtr? envHandle = null)
```
Where there is default parameters instead of overloading.

**Can C# language version be updated? To hopefully at least 8.**
(Null coalescing assignment ❤️ )

For more background see issue:
#13743 

## Tasks

- [ ] will the marshalling handle bool to int?
- [ ]  add `OrtEnv` as a parameter to `InferenceSession` ctors
- [ ] `DisablePerSessionThreads()` implicit when using `GetEnvironment(opts)`?

### CODE REVIEW

**Comment by thoron** [comment]
> @thoron please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> 
> Contributor License Agreement

@microsoft-github-policy-service agree [company="Prediktera AB"]

**Comment by thoron** [comment]
@microsoft-github-policy-service agree company="Prediktera AB"

**Comment by thoron** [comment]
> @thoron please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> 
> Contributor License Agreement

@microsoft-github-policy-service agree company="Prediktera AB"

File: csharp/src/Microsoft.ML.OnnxRuntime/OrtEnv.shared.cs [hunk 1]
```diff
@@ -0,0 +1,191 @@
+﻿using System;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.ML.OnnxRuntime
+{
+    /// <summary>
+    /// This class handles the process-global ONNX Runtime environment instance (OrtEnv)
+    /// </summary>
+    public sealed class OrtEnv : SafeHandle
```
Comment (yuslepukhin) [inline csharp/src/Microsoft.ML.OnnxRuntime/OrtEnv.shared.cs:9]: >public sealed class OrtEnv : SafeHandle [](http://example.com/codeflow?start=4&length=39)

OrtEnv class already exists in the C# API. You may want to add new functionality there.

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests.NetCoreApp/InferenceTest.netcore.cs [hunk 2]
```diff
@@ -686,10 +686,13 @@ private void TestRegisterCustomOpLibrary()
                 string libFullPath = Path.Combine(Directory.GetCurrentDirectory(), libName);
                 Assert.True(File.Exists(libFullPath), $"Expected lib {libFullPath} does not exist.");
 
-                var ortEnvInstance = OrtEnv.Instance();
-                string[] providers = ortEnvInstance.GetAvailableProviders();
-                if (Array.Exists(providers, provider => provider == "CUDAExecutionProvider")) {
-                    option.AppendExecutionProvider_CUDA(0);
+                using (var ortEnvInstance = OrtEnv.Instance())
```
Comment (yuslepukhin) [inline csharp/test/Microsoft.ML.OnnxRuntime.Tests.NetCoreApp/InferenceTest.netcore.cs:689]: > using (var ortEnvInstance = OrtEnv.Instance()) [](http://example.com/codeflow?start=15&length=47)

It is a singleton, we never release it.

**Review by yuslepukhin — CHANGES_REQUESTED** [review]
:clock1:

File: csharp/test/Microsoft.ML.OnnxRuntime.Tests.NetCoreApp/InferenceTest.netcore.cs (hunk 2 above)
Comment (thoron) [reply csharp/test/Microsoft.ML.OnnxRuntime.Tests.NetCoreApp/InferenceTest.netcore.cs:689]: I added the scoping of the `OrtEnv` in the tests to be able to create new instances using `ThreadingOptions`. Normally the class would be used as a singleton and throwing exceptions when attempting to create a new instance using `ThreadingOptions`.

I added a unit test to highlight this feature:

https://github.com/thoron/onnxruntime/blob/e30d55f9b51e44557b8c535f8a006f639d585037/csharp/test/Microsoft.ML.OnnxRuntime.Tests.Common/InferenceTest.cs#L273

File: csharp/src/Microsoft.ML.OnnxRuntime/OrtEnv.shared.cs (hunk 1 above)
Comment (thoron) [reply csharp/src/Microsoft.ML.OnnxRuntime/OrtEnv.shared.cs:9]: Yes, I moved the class from being in this file to its own class file (I thought this conformed more to the rest of the project). I now see that the file `OnnxRuntime.shared.cs` does not contain a lot after i migrated `OrtEnv` to its own file. Should i move the code back from `OrtEnv.shared.cs` to `OnnxRuntime.shared.cs` and rename it?

- [ ] Rename `OnnxRuntime.shared.cs`?
- [ ] Remove `OrtEnv.shared.cs` ?

Comment (yuslepukhin) [reply csharp/src/Microsoft.ML.OnnxRuntime/OrtEnv.shared.cs:9]: It makes it harder to review your specific change. I do not have an objection at this time on moving in a separate PR.

File: csharp/src/Microsoft.ML.OnnxRuntime/ThreadingOptions.Shared.cs [hunk 3]
```diff
@@ -0,0 +1,67 @@
+﻿using System;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.ML.OnnxRuntime
+{
+    public class ThreadingOptions : SafeHandle
```
Comment (yuslepukhin) [inline csharp/src/Microsoft.ML.OnnxRuntime/ThreadingOptions.Shared.cs:6]: > public class ThreadingOptions : SafeHandle [](http://example.com/codeflow?start=3&length=43)

This will fail documentation check. Please, refer to other classes. At a very least every class and every public member must be documented.

Comment (yuslepukhin) [reply csharp/src/Microsoft.ML.OnnxRuntime/ThreadingOptions.Shared.cs:6]: To run the documentation check locally you would need to install doxygen and run it similar to what is in our pipelines:

Point ORT_DOXY_SRC to the location of the source code.
Point ORT_DOXY_OUT the location of the doxygen output folder.

set ORT_DOXY_SRC=$(Build.SourcesDirectory)
set ORT_DOXY_OUT=$(Build.BinariesDirectory)\${{ parameters.BuildConfig }}\${{ parameters.BuildConfig }}
mkdir %ORT_DOXY_SRC%
 mkdir %ORT_DOXY_OUT%
     "C:\Program Files\doxygen\bin\doxygen.exe" $(Build.SourcesDirectory)\tools\ci_build\github\Doxyfile_csharp.cfg

File: csharp/src/Microsoft.ML.OnnxRuntime/OnnxRuntime.shared.cs [hunk 4]
```diff
@@ -90,8 +103,29 @@ internal static IntPtr Handle
         /// It returns the same instance on every call - `OrtEnv` is singleton
         /// </summary>
         /// <returns>Returns a singleton instance of OrtEnv that represents native OrtEnv object</returns>
-        public static OrtEnv Instance() { return _instance.Value; }
+        public static OrtEnv Instance()
+        {
+            return _instance ?? (_instance = new OrtEnv());
```
Comment (yuslepukhin) [inline csharp/src/Microsoft.ML.OnnxRuntime/OnnxRuntime.shared.cs:108]: >return _instance ?? (_instance = new OrtEnv()); [](http://example.com/codeflow?start=12&length=47)

This is not thread-safe. Lazy<> mechanism was used there for good reasons. One of them is thread-safety.

Comment (thoron) [reply csharp/src/Microsoft.ML.OnnxRuntime/OnnxRuntime.shared.cs:108]: I would suggest removing the `Instance()` method fully replacing it with the same structure the java code has:
https://javadoc.io/doc/com.microsoft.onnxruntime/onnxruntime/latest/ai/onnxruntime/OrtEnvironment.html
Only exposing `GetEnvironment` methods. This however may result in some regression.

I can just add a simple `lock` to ensure thread safety.

- [ ] Remove `SafeHandle`?
- [ ] Add `lock` to instantiations
- [ ] Remove `Instance()` method?

What should be done in the scope of this PR?

(Side note, the java version also implements `AutoCloseable` which would have the same possible limitations as `SafeHandle`)

Comment (yuslepukhin) [reply csharp/src/Microsoft.ML.OnnxRuntime/OnnxRuntime.shared.cs:108]: Here is how I see the obstacle. We are trying to modify the implementation for the sake of testing in a way that is not acceptable. Here is my suggestion.
Create a `SafeHandle` implementation class that would actually instantiate and destroy the native handle as before. 

The `Env` class would use it to instantiate the native instance and retrieve the handle. It would dispose the implementation instance at the right time. The `Env` class we can simply make `Disposable` instead of `SafeHandle`.

Now, you can have a class that you test directly in your tests for the functionality you want to implement without changing the API or runtime behavior of `Env`.

There are details you will have to iron out. For example, make the class accessible to the test code and some others.

File: csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.shared.cs [hunk 5]
```diff
@@ -136,6 +136,17 @@ public InferenceSession(byte[] model, SessionOptions options)
         {
             Init(model, options);
         }
+        
+        /// <summary>
```
Comment (yuslepukhin) [inline csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.shared.cs:140]: I feel this should not be a part of the change. Let' make another PR

File: csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.shared.cs [hunk 6]
```diff
@@ -136,6 +136,17 @@ public InferenceSession(byte[] model, SessionOptions options)
...
+        /// Constructs an InferenceSession from a model data in byte array, with some additional session options
+        /// </summary>
+        /// <param name="model"></param>
+        /// <param name="options"></param>
+        /// <param name="env"></param>
+        public InferenceSession(byte[] model, SessionOptions options, OrtEnv env)
+        {
+            Init(model, options, null, env.DangerousGetHandle());
```
Comment (yuslepukhin) [inline csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.shared.cs:148]: The convention is to use Handle

File: csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.shared.cs [hunk 7]
```diff
@@ -766,43 +777,49 @@ public ulong ProfilingStartTimeNs
         #region private methods
 
         private void Init(string modelPath, SessionOptions options,
-                          PrePackedWeightsContainer prepackedWeightsContainer = null)
+                          PrePackedWeightsContainer prepackedWeightsContainer = null, IntPtr? envHandle = null)
```
Comment (yuslepukhin) [inline csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.shared.cs:780]: Let's focus on the threading options.

Comment (yuslepukhin) [inline csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.shared.cs:780]: We confine native handles to the code that invokes Native API. Otherwise, we do not pass them around.

Comment (yuslepukhin) [inline csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.shared.cs:780]: It would be better to pass the env object. We confine native handles to the code that actually invokes the native API.

**Comment by yuslepukhin** [comment]
One thing that is missing in the PR description is why Environment level global threadpool is interesting? I think it would be easier to configure session level threadpools via SessionOptions.

File: csharp/src/Microsoft.ML.OnnxRuntime/OnnxRuntime.shared.cs (hunk 4 above)
Comment (thoron) [reply csharp/src/Microsoft.ML.OnnxRuntime/OnnxRuntime.shared.cs:108]: It seems as a quite big rewrite of the current `OrtEnv` class, I am a bit hesitant to do this. 

I could add an `internal` class in OrtEnv for the explicit purpose of destroying the singleton instance.
```csharp
internal class OrtEnvInstanceScoper : IDisposable
{
    public OrtEnv Env { get; }

    public OrtEnvInstanceScoper(ThreadingOptions threadingOptions = null)
    {
        Env = threadingOptions == null ? Instance() : GetEnvironment(threadingOptions);
    }
    public void Dispose()
    {
        Env.ReleaseHandle();
        Env.ClearInstance(); // private
    }
}
```

File: csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.shared.cs (hunk 5 above)
Comment (thoron) [reply csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.shared.cs:140]: I keep the ctor but made it `internal` so it still can be part of the unit test suit. Will expand on the constructors in another PR.

**Comment by thoron** [comment]
@yuslepukhin I screwed up with the merge, sorry! Tried to use some builtin github tools in the IDE and did not realize which actions were taken. I will try to fix it. Just ignore for now, please.

**Comment by yuslepukhin** [comment]
Please, keep this PR around, I am still pondering on this. Thanks!

**Comment by thoron** [comment]
> Please, keep this PR around, I am still pondering on this. Thanks!

I will close this PR and in favor for yours #14723
